api/client/fpa.go

```diff
@@ -32,6 +32,42 @@ func (f *fpa) Store(ctx context.Context, c cid.Cid) error {
 	return nil
 }
 
+func (f *fpa) AddFile(ctx context.Context, data io.Reader) (*cid.Cid, error) {
+	stream, err := f.client.AddFile(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer := make([]byte, 1024*32) // 32KB
+	for {
+		bytesRead, err := data.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		sendErr := stream.Send(&pb.AddFileRequest{Chunk: buffer[:bytesRead]})
+		if sendErr != nil {
+			if sendErr == io.EOF {
+				var noOp interface{}
+				return nil, stream.RecvMsg(noOp)
```

Comment:
Oh, this is important. I was thrown off by this for 24 hrs. Turns out that if the server returns an error in the middle of client streaming requests, you get back error `io.EOF`. If you want to see the actual error from the server, you have to get it with `stream.RecvMsg()`. Will get this updated elsewhere in the code base.

Reply:
Good catch!